File: advent-of-code-2023/dart/bin/day02.dart
import 'dart:io';
import 'dart:math';

import 'challenge_day.dart';

class Day02 implements ChallengeDay {
  final List<String> _lines;

  Day02(this._lines);

  Day02.using(String fileName) : _lines = File(fileName).readAsLinesSync();

  @override
  int part1() => _lines
      .map((line) => Game.parse(line))
      .where((game) => game.isPossible())
      .fold(0, (sum, game) => sum + game._id);

  @override
  int part2() => _lines
      .map((line) => Game.parse(line))
      .fold(0, (sum, game) => sum + game.toPowerFewestCubes());

  @override
  String message() => 'Day 02: ';
}

class Game {
  final int _id;
  final List<String> _rounds;

  Game(this._id, this._rounds);

  int toPowerFewestCubes() {
    int redCount = 1;
    int blueCount = 1;
    int greenCount = 1;
    for (var round in _rounds) {
      final List<List<String>> actions = toGameActions(round);
      for (final List<String> action in actions) {
        final int count = int.parse(action.first);
        final String color = action.last;
        switch (color) {
          case "red":
            redCount = max(count, redCount);
          case "green":
            greenCount = max(count, greenCount);
          case "blue":
            blueCount = max(count, blueCount);
        }
      }
    }
    return redCount * blueCount * greenCount;
  }

  bool isPossible() {
    for (var round in _rounds) {
      var redCount = 0;
      var blueCount = 0;
      var greenCount = 0;
      final List<List<String>> actions = toGameActions(round);
      for (var action in actions) {
        final int count = int.parse(action.first);
        final String color = action.last;
        switch (color) {
          case "red":
            redCount += count;
          case "green":
            greenCount += count;
          case "blue":
            blueCount += count;
        }
        if (redCount > 12 || greenCount > 13 || blueCount > 14) {
          return false;
        }
      }
    }
    return true;
  }

  static List<List<String>> toGameActions(String round) =>
      round.split(", ").map((a) => a.split(" ")).toList();

  static Game parse(String line) {
    final List<String> split = line.split(": ");
    final int id = int.parse(split.first.substring("Game ".length));
    final List<String> subSets = split.last.split("; ").toList();
    return Game(id, subSets);
  }
}

File: advent-of-code-2024/src/main/java/aoc/snowrescuemission/Day02.java
package aoc.snowrescuemission;

import aoc.utils.ChallengeDay;
import org.hzt.utils.collections.ListX;
import org.hzt.utils.io.FileX;
import org.hzt.utils.strings.StringX;

import java.util.List;

import static java.lang.Math.max;

public final class Day02 implements ChallengeDay {

    private final ListX<Game> games;

    public Day02(String fileName) {
        this.games = FileX.of(fileName).useLines(s -> s.map(Game::parse).toListX());
    }

    public Day02(ListX<String> lines) {
        this.games = lines.map(Game::parse);
    }

    @Override
    public Long part1() {
        return games
                .filter(Game::isPossible)
                .longSumOf(Game::id);
    }

    @Override
    public Long part2() {
        return games.longSumOf(Game::toPowerFewestCubes);
    }

    record Game(int id, List<String> rounds) {

        public int toPowerFewestCubes() {
            int redCount = 1;
            int blueCount = 1;
            int greenCount = 1;
            for (final var round : rounds) {
                final var actions = toGameActions(round);
                for (final var action : actions) {
                    final var count = Integer.parseInt(action.first());
                    final var color = action.last();
                    switch (color) {
                        case "red" -> redCount = max(count, redCount);
                        case "green" -> greenCount = max(count, greenCount);
                        case "blue" -> blueCount = max(count, blueCount);
                    }
                }
            }
            return redCount * blueCount * greenCount;
        }

        public boolean isPossible() {
            for (final var round : rounds) {
                int redCount = 0;
                int blueCount = 0;
                int greenCount = 0;
                final var actions = toGameActions(round);
                for (final var action : actions) {
                    final var count = Integer.parseInt(action.first());
                    final var color = action.last();
                    switch (color) {
                        case "red" -> redCount += count;
                        case "green" -> greenCount += count;
                        case "blue" -> blueCount += count;
                    }
                    if (redCount > 12 || greenCount > 13 || blueCount > 14) {
                        return false;
                    }
                }
            }
            return true;
        }

        private static ListX<ListX<String>> toGameActions(String round) {
            return StringX.of(round).split(", ")
                    .map(a -> StringX.of(a).split(" "));
        }

        private static Game parse(String line) {
            final var split = StringX.of(line).split(": ");
            final var id = Integer.parseInt(split.first().substring("Game ".length()));
            final var subSets = StringX.of(split.last()).split("; ").toList();

            return new Game(id, subSets);
        }
    }
}
